### DISCUSSION TITLE
Steps for custom role in Jupyter Book?

### DISCUSSION THREAD

**Discussion by matthew-brett** [executablebooks/meta#101] [Q&A] [answered]
I have a custom role, in fact called `data8page`, with the usual Sphinx boilerplate, and returning a paragraph node.

It works fine in a tiny sphinx project, producing the rendered paragraph text.

However, when I use it in some Markdown in JupyterBook, the role does not get parsed, so the input text remains unaltered in the rendered source:

```
{data8page}`foo`
```

Standard Sphinx roles such as `{abbr}` and `{math}` work correctly.

I am importing the role in `_config.yml`:

```
sphinx:
  extra_extensions:
    - cfdcode.data8page
```

in the same form as I'm using in a standalone `conf.py` for a Sphinx project:

```
extensions = ['cfdcode.data8page']
```

Am I doing this correctly?  Any ideas how I could debug why it is not being correctly parsed / rendered?

**Comment by chrisjsewell** [comment]
Heya, so is there then a warning shown that the `data8page` role does not exist?

**Comment by matthew-brett** [comment]
There's nothing in the stdout / stderr from `jupyter-book build .` about `data8page`.  Should I be looking somewhere else?

**Reply by choldgraf** [reply]
that's weird...if you put an error in the role code does anything happen?

**Reply by matthew-brett** [reply]
Ah - sorry - I should have said.   If I put `1 / 0` into the role function, then no, it does *not* error, so the role function is not being called.   I do get an error from the `setup` function, that initializes the role, when I put `1 / 0` into that.

**Comment by choldgraf** [comment]
can you provide a minimum book repo that recreates the problem? It's weird that the setup function is called but the role code isn't being called!

**Reply by matthew-brett** [reply]
It is weird!  

Here's the repo - see the README for instructions : https://github.com/matthew-brett/jbtest

**Comment by choldgraf** [comment]
So, I think this might be a myst parser thing as opposed to a jupyter book thing.

In your mini Sphinx site, if I activate `myst_parser` and add a markdown file with this content:

```md
{math}`a = b + c`

{data8page}`some text`

{blah}`foo`
```

then I see this output:

![image](https://user-images.githubusercontent.com/1839645/90322648-a0956100-df0b-11ea-9cdf-7dc86503e7ee.png)

The rST, this:

```rst
:math:`a = b + c`

:data8page:`some text`

:blah:`foo`
```

yields this:

![image](https://user-images.githubusercontent.com/1839645/90322652-af7c1380-df0b-11ea-8658-3d7813bc5c06.png)

Note that none of this is jupyter book, just raw sphinx, and note that the `data8page` role is being treated *differently* from an unregistered role, it is just not putting out the right output

**Comment by matthew-brett** [comment]
Interesting.

I see that, in this repo, if I put at `1 / 0` error in the role function, it does error in the Jupyter Book build, so I guess it is being called.  I was pretty sure I'd checked that thoroughly before, maybe I was wrong, or was decieved by caching.

**Comment by choldgraf** [comment, marked as answer]
further sleuthing that I suspect @chrisjsewell has figured out as well (see below)

if I do:

```python
def setup(app):
    # Add runrole roles
    app.add_role('datapage', data8pagerole)
    app.add_role('data8page', data8pagerole)
```

and add this to a page

```md
{data8page}`some text`

{datapage}`some text`
```

then the `datapage` one shows up as expected and the `data8page` does not. So I suspect that this is something about there being a number in your role name.

I further suspect this because I see @chrisjsewell just opened up https://github.com/executablebooks/markdown-it-py/pull/34 which adds numbers to the regex in the role search :-)

**Reply by chrisjsewell** [reply]
your sleuthing knows no limits 🧐 but yeh I beat you to it by a couple of hours lol

**Reply by choldgraf** [reply]
haha

Chris H: discover a problem and then pedantically tell people about it
Chris S: discover a problem and actually fix it

😅

**Reply by chrisjsewell** [reply]
I'm just playing around with some other bits, then I'll merge and release soon

**Comment by westurner** [comment]
"How to use roles and directives" https://github.com/executablebooks/MyST-NB/issues/239

> - [ ] Show examples of how to use roles and directives (with arguments)
